web/src/main/java/org/mskcc/oncotree/api/TumorTypesApi.java

```diff
@@ -181,42 +101,36 @@
             }*/
             tumorTypes.add(matchedTumorTypes);
         }
-
-        resp.setData(tumorTypes);
-        return new ResponseEntity<SearchTumorTypesPostResp>(resp, HttpStatus.OK);
+        return tumorTypes;
     }
 
 
-    @ApiOperation(value = "Tumor Types", notes = "", response = SearchTumorTypesResp.class)
-    @io.swagger.annotations.ApiResponses(value = {
-        @io.swagger.annotations.ApiResponse(code = 200, message = "An array of tumor types")})
+    @ApiOperation(value = "Tumor Types", notes = "", response = TumorType.class)
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = "An array of tumor types")})
     @RequestMapping(value = "/search/{type}/{query}",
-        produces = {"application/json"},
+        produces = {APPLICATION_JSON_VALUE},
         method = RequestMethod.GET)
-    public ResponseEntity<SearchTumorTypesResp> tumorTypesSearchTypeQueryQueryGet(
+    public Iterable<TumorType> tumorTypesSearchTypeQueryQueryGet(
         @ApiParam(value = "Query type. It could be 'id', 'code', 'name', 'mainType', 'level', 'nci', 'umls' or 'color'. You can also use 'all' to search all content.", required = true)
         @PathVariable("type") String type,
         @ApiParam(value = "The query content", required = true)
         @PathVariable("query") String query,
-        @ApiParam(value = "The version of tumor types. For example, 1, 1.1 Please see GitHub for released versions. ")
+        @ApiParam(value = "The version of tumor types. For example, " + VersionUtil.DEFAULT_VERSION + ". Please see the versions api documentation for released versions.")
         @RequestParam(value = "version", required = false) String version,
         @ApiParam(value = "If it sets to true, it will only return one element array.", defaultValue = "true")
         @RequestParam(value = "exactMatch", required = false, defaultValue = "true") Boolean exactMatch,
         @ApiParam(value = "Tumor type levels. 1-5. By default, it doesn't includes tissue which is the primary level.", defaultValue = "2,3,4,5")
         @RequestParam(value = "levels", required = false, defaultValue = "1,2,3,4,5") String levels
-//        , @ApiParam(value = "The callback function name. This has to be used with dataType JSONP.")
-//        @RequestParam(value = "callback", required = false) String callback
-    )
-        throws NotFoundException {
+    ) {
         List<TumorType> matchedTumorTypes = new ArrayList<>();
         Version v = (version == null) ? VersionUtil.getDefaultVersion() : VersionUtil.getVersion(version);
 
         // Cache in tumor types in case no data present
-        CacheUtil.getOrResetTumorTypesByVersion(v);
+        CacheUtil.getTumorTypesByVersion(v);
 
         matchedTumorTypes = v == null ? new ArrayList<TumorType>() : TumorTypesUtil.findTumorTypesByVersion(type, query, exactMatch, v, false);
-        SearchTumorTypesResp resp = new SearchTumorTypesResp();
-
+        // TODO add back or remove levels option
```

Comment:
Also here, we should delete this or add it back and fix it.

Reply:
I do like the levels. It is quite useful when you have a nested object.

Reply:
I also like the idea of specifying levels, but we should talk through the behavior and understand the design/requirements. I think that the search api endpoints always return a flat response (not nested). In this case, specifying the levels could simply act as a filter, removing all non-matching nodes from the list. If the /api/tumorTypes/search endpoint could return a nested representation (similar to the /api/tumorTypes endpoint) then I think we would have difficulty representing the child-parent relationship in the response. The field "parent" is not too much of a problem, since it is simply a string. But the field "children" is a fully populated object of type TumorType ... and if a user requested levels "2,4" then I wonder whether the level 4 nodes would appear as children of level 2? Somehow that seems incorrect. If we wanted a nested representation which restricted the levels returned we might need to change to a "filter all levels above" / "filter all levels below" approach so that there are no gap levels in between parents and children. Also, there is a question of how to contain the nodes in the uppermost (lowest numbered) level. In the GET /api/tumorTypes endpoint there is one root node (TISSUE) at level 0 which contains as children all the level 1 nodes (the tissues). If we returned a nested response from search and filtered all levels from 0 through 2 then I guess the response would be a list of nested trees, each one starting at level 3 ?